ingest/ingest_retry_client.go

```diff
@@ -0,0 +1,147 @@
+func NewRetryClient(ctx context.Context, retryPolicy *RetryPolicy, opts ...api.ClientOption) (*RetryClient, error) {
+	client, err := NewClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	c := &RetryClient{
+		Client:      client,
+		retryPolicy: retryPolicy,
+	}
+
+	if _, err = time.ParseDuration(c.retryPolicy.InitialBackoff); err != nil {
+		return nil, errors.New(codes.FailedPrecondition, fmt.Sprintf("invalid backoff duration: %v", err))
+	}
+	return c, nil
+}
+
+// NewTestRetryClient creates a new Ingest API gRPC Client with retry functionality.
+func NewTestRetryClient(client ingestpb.IngestAPIClient, retryPolicy *RetryPolicy) (*RetryClient, error) {
+	c, err := NewTestClient(client)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	rc := &RetryClient{
+		Client:      c,
+		retryPolicy: retryPolicy,
+	}
+
+	if _, err = time.ParseDuration(rc.retryPolicy.InitialBackoff); err != nil {
+		return nil, errors.New(codes.FailedPrecondition, fmt.Sprintf("invalid backoff duration: %v", err))
+	}
+	return rc, nil
+}
+
+// SendRecord sends a record on the opened stream and retries if it fails.
+func (c *RetryClient) SendRecord(record *ingestpb.Record) error {
+	if c.stream == nil {
+		return errors.New(codes.FailedPrecondition, "a stream must be opened first")
+	}
+
+	err := c.Client.SendRecord(record)
+	if err == nil {
+		return err
+	}
+	return c.sendRecordWithRetry(record)
+}
+
+// ReceiveResponse returns the next response available on the stream.
+func (c *RetryClient) ReceiveResponse() (*ingestpb.StreamRecordsResponse, error) {
+	if c.stream == nil {
+		return nil, errors.New(codes.FailedPrecondition, "a stream must be opened first")
+	}
+
+	resp, err := c.Client.ReceiveResponse()
+	if err == nil {
+		return resp, err
+	}
+	return c.receiveResponseWhenReconnected()
+}
+
+// receiveResponseWhenReconnected returns try to read the next response available on the stream
+// or return error if unable to reconnect.
+func (c *RetryClient) receiveResponseWhenReconnected() (*ingestpb.StreamRecordsResponse, error) {
+	for {
+		if c.isUnableToReconnect {
+			return nil, errors.New(codes.Unavailable, "unable to reconnect to server")
+		}
+		time.Sleep(1 * time.Second)
+		resp, err := c.Client.ReceiveResponse()
+		if err == nil {
+			return resp, err
+		}
+	}
+}
+
+// sendRecordWithRetry creates a new stream and tries sends a record. Will retry based on retry policy.
+func (c *RetryClient) sendRecordWithRetry(record *ingestpb.Record) error {
+	backoffTime, _ := time.ParseDuration(c.retryPolicy.InitialBackoff)
+	var err error
+	for i := 1; i <= c.retryPolicy.MaxAttempts; i++ {
+		log.Printf("attempting to reconnect (%d/%d) in %s...", i, c.retryPolicy.MaxAttempts, backoffTime.String())
+		time.Sleep(backoffTime)
+		backoffTime *= time.Duration(c.retryPolicy.BackoffMultiplier)
+		ctx, _ := context.WithTimeout(context.Background(), c.retryPolicy.CtxTimeout) //nolint
```

Comment:
I'm unsure about this one, linter complained if I did not use and call the cancel func.

Reply:
Yes, but the linter still complains because the cancel func is in a for loop. I've also tried to put the declaration of the func outside the for loop, but still lint issues. 🤷‍♂️

Reply:
Here is nice explanation: https://stackoverflow.com/a/45620423

TLDR: wrap the content of the loop in a function (either anonymous or named one)

Reply:
Context issue solve!